### PR TITLE
feat: add support of multiple templates for BGD cases

### LINKED ISSUE #851: [Feat]: Separate template artifacts for origin and peer namespaces
### Is your feature request related to a problem? Please describe

_No response_

### Describe the solution you'd like

To support Blue-Green Deployment (BGD) with separate template artifacts for origin and peer namespaces, the following requirements must be met:

- [ ] The Environment Inventory must support optional `envTemplate.bgNsArtifacts.origin` and `envTemplate.bgNsArtifacts.peer` attributes according to the [description](https://github.com/Netcracker/qubership-envgene/blob/main/docs/features/environment-instance-generation.md#template-artifacts)
  - [Use Cases](https://github.com/Netcracker/qubership-envgene/blob/main/docs/use-cases/environment-instance-generation.md#template-artifacts-selection) 

- [ ] The Instance pipeline must support [`ENV_TEMPLATE_VERSION_ORIGIN`](https://github.com/Netcracker/qubership-envgene/blob/main/docs/instance-pipeline-parameters.md#env_template_version_origin), [`ENV_TEMPLATE_VERSION_PEER`](https://github.com/Netcracker/qubership-envgene/blob/main/docs/instance-pipeline-parameters.md#env_template_version_peer) parameters to update `envTemplate.bgNsArtifacts.origin`, `envTemplate.bgNsArtifacts.peer` in the Environment Inventory according

- [ ] During the `warmup` operation in the `bg_manage` job , the system must copy `envTemplate.bgNsArtifacts.origin/peer` → `envTemplate.bgNsArtifacts.peer/origin` in the Environment Inventory according to the [description](https://github.com/Netcracker/qubership-envgene/blob/main/docs/features/blue-green-deployment.md#warmup-operation)
  - [Use Cases](https://github.com/Netcracker/qubership-envgene/blob/main/docs/use-cases/environment-instance-generation.md#namespace-folder-name-generation)

Documentation branch: https://github.com/Netcracker/qubership-envgene/tree/main

**Automated testing must be implemented**

### Describe alternatives you've considered

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by GlimmerCape**
# Pull Request

## Summary

Provide a concise description of what this pull request does and why it is needed.

## Issue

Closes #851

## Breaking Change?

- [ ] Yes
- [ ] No

If yes, describe the breaking change and its impact (e.g., API changes, behavior changes, or required updates for users).

## Scope / Project

Specify the component, module, or project area affected by this change (e.g., `docs`, `actions`, `workflows`).

## Implementation Notes

Provide details on how the change was implemented, including any technical considerations, trade-offs, or notable design decisions. Leave blank if not applicable.

## Tests / Evidence

Describe how the changes were verified, including:

- Tests added or updated (e.g., unit, integration, end-to-end)
- Manual testing steps or results
- Screenshots, logs, or other evidence (if applicable)

## Additional Notes

Include any extra information, such as:

- Dependencies introduced
- Future work or follow-up tasks
- Reviewer instructions or context
- References to related PRs or discussions

Leave blank if not applicable.

### CODE REVIEW

File: scripts/build_env/main.py [hunk 1]
```diff
@@ -28,16 +28,22 @@ def prepare_folders_for_rendering(env_name, cluster_name, source_env_dir, templa
     # clearing instances dir
     cleanup_resulting_dir(Path(output_dir) / cluster_name / env_name)
     # copying parameters from templates and instances
-    check_dir_exist_and_create(f'{render_parameters_dir}/from_template')
-    copy_path(f'{templates_dir}/parameters', f'{render_parameters_dir}/from_template')
+    for k, v in templates_dirs.items():
```
Comment (miyamuraga) [inline scripts/build_env/main.py]: not good readable. it is not necessary to name **letters** that are not immediately obvious from code

File: scripts/build_env/main.py [hunk 2]
```diff
@@ -288,11 +296,19 @@ def render_environment(env_name, cluster_name, templates_dir, all_instances_dir,
     base_dir = getenv_with_error('CI_PROJECT_DIR')
     cluster = getenv_with_error("CLUSTER_NAME")
     environment = getenv_with_error("ENVIRONMENT_NAME")
-    g_templates_dir = f"{base_dir}/tmp/templates"
+    g_template_dirs = {
+        'common': f"{base_dir}/tmp/templates",
+    }
+    origin_template_path = f"{base_dir}/tmp/origin/templates"
```
Comment (miyamuraga) [inline scripts/build_env/main.py]: It is better to make constant in helper for repetitive static path

File: scripts/build_env/main.py [hunk 3]
```diff
@@ -288,11 +296,19 @@ def render_environment(env_name, cluster_name, templates_dir, all_instances_dir,
...
+    if check_dir_exists(origin_template_path):
+        g_template_dirs['origin'] = origin_template_path
```
Comment (miyamuraga) [inline scripts/build_env/main.py]: why not use enum for types like peer, origin, controller?

File: scripts/build_env/main.py [hunk 4]
```diff
@@ -160,7 +166,8 @@ def build_environment(env_name, cluster_name, templates_dir, source_env_dir, all
     envvars["env"] = env_name  # Keep as string for file paths
     envvars["current_env"] = current_env  # Object for Jinja2 templates that need current_env.environmentName
     envvars["cluster_name"] = cluster_name
-    envvars["templates_dir"] = templates_dir
+    envvars["templates_dirs"] = templates_dirs
+    envvars["templates_dir"] = templates_dirs.get('common', '')
```
Comment (miyamuraga) [inline scripts/build_env/main.py]: why do we pass templates_dir if it is part of templates_dirs?

File: scripts/build_env/render_config_env.py [hunk 5]
```diff
@@ -23,12 +23,15 @@ class Context(BaseModel):
     render_dir: Optional[str] = ''
     cloud_passport: OrderedDict = Field(default_factory=OrderedDict)
     templates_dir: Optional[Path] = None
```
Comment (miyamuraga) [inline scripts/build_env/render_config_env.py]: need to delete

File: python/envgene/envgenehelper/business_helper.py [hunk 6]
```diff
@@ -367,16 +370,36 @@ def find_cloud_name_from_passport(source_env_dir, all_instances_dir):
     else:
         return ""
 
+class NamespaceRole(Enum):
+    COMMON = 1
```
Comment (miyamuraga) [inline python/envgene/envgenehelper/business_helper.py]: what sense of numbers in this case

File: scripts/build_env/build_env.py [hunk 7]
```diff
@@ -38,6 +46,28 @@ def processFileList(mask, dict, dirPointer):
             dict[extractNameFromFile(filePath)] = [{"filePath": filePath, "envSpecific": False}]
     return dict
 
+def create_role_specific_paramset_map(base_paramset_map: dict, role: NamespaceRole,
+                                      origin_template_exists: bool, peer_template_exists: bool) -> dict:
+    if role == NamespaceRole.ORIGIN and origin_template_exists:
+        excluded_dirs = ['from_template', 'from_peer_template']
```
Comment (miyamuraga) [inline scripts/build_env/build_env.py]: the same, must be constants

File: scripts/build_env/build_env.py [hunk 8]
```diff
@@ -38,6 +46,28 @@ def processFileList(mask, dict, dirPointer):
...
+    elif role == NamespaceRole.PEER and peer_template_exists:
+        excluded_dirs = ['from_template', 'from_origin_template']
+    else:
+        excluded_dirs = ['from_origin_template', 'from_peer_template']
+
+    result = {}
+
+    for paramset_name, entries in base_paramset_map.items():
+        kept = [
```
Comment (miyamuraga) [inline scripts/build_env/build_env.py]: bad naming

File: scripts/build_env/build_env.py [hunk 9]
```diff
@@ -38,6 +46,28 @@ def processFileList(mask, dict, dirPointer):
...
+    if role == NamespaceRole.ORIGIN and origin_template_exists:
+        excluded_dirs = ['from_template', 'from_peer_template']
+    elif role == NamespaceRole.PEER and peer_template_exists:
+        excluded_dirs = ['from_template', 'from_origin_template']
+    else:
+        excluded_dirs = ['from_origin_template', 'from_peer_template']
+
+    result = {}
+
+    for paramset_name, entries in base_paramset_map.items():
```
Comment (miyamuraga) [inline scripts/build_env/build_env.py]: bad naming

File: scripts/build_env/build_env.py [hunk 10]
```diff
@@ -38,6 +46,28 @@ def processFileList(mask, dict, dirPointer):
             dict[extractNameFromFile(filePath)] = [{"filePath": filePath, "envSpecific": False}]
     return dict
 
+def create_role_specific_paramset_map(base_paramset_map: dict, role: NamespaceRole,
```
Comment (miyamuraga) [inline scripts/build_env/build_env.py]: it's not immediately obvious what the key is in map and what the value is

File: scripts/build_env/build_env.py [hunk 11]
```diff
@@ -523,22 +559,33 @@ def build_env(env_name, env_instances_dir, parameters_dir, env_template_dir, res
...
-    for templatePath in namespaceTemplates:
-        logger.info(f"Processing namespace: {templatePath}")
-        templateName = getTemplateNameFromNamespacePath(templatePath)
-        template_namespace_names.append(templateName)
-        initParametersStructure(env_specific_parameters_map["namespaces"], templateName)
+    origin_paramset_map = create_role_specific_paramset_map(paramset_map, NamespaceRole.ORIGIN,
+                                                            origin_template_exists, peer_template_exists)
+    peer_paramset_map = create_role_specific_paramset_map(paramset_map, NamespaceRole.PEER,
+                                                          origin_template_exists, peer_template_exists)
+    common_paramset_map = create_role_specific_paramset_map(paramset_map, NamespaceRole.COMMON,
```
Comment (miyamuraga) [inline scripts/build_env/build_env.py]: if not peer and not origin will return the same map for origin_paramset_map. peer_paramset_map, common_paramset_map. unnecessary method calls. need fix logic of create_role_specific_paramset_map

File: scripts/build_env/build_env.py [hunk 12]
```diff
@@ -5,15 +5,23 @@
 import yaml
 import re
 import pathlib
+from pathlib import Path
+
 from envgenehelper import *
+from envgenehelper.business_helper import get_namespace_role, get_namespaces, NamespaceRole
```
Comment (miyamuraga) [inline scripts/build_env/build_env.py]: delete imports that are not used

File: scripts/build_env/build_env.py [hunk 13]
```diff
@@ -523,22 +559,33 @@ def build_env(env_name, env_instances_dir, parameters_dir, env_template_dir, res
 
     # process namespaces
     template_namespace_names = []
-    # iterate through namespace definitions and create namespace parameters
-    for templatePath in namespaceTemplates:
-        logger.info(f"Processing namespace: {templatePath}")
-        templateName = getTemplateNameFromNamespacePath(templatePath)
```
Comment (miyamuraga) [inline scripts/build_env/build_env.py:529]: need to delete useless method getTemplateNameFromNamespacePath also

File: scripts/build_env/build_env.py [hunk 14]
```diff
@@ -5,15 +5,23 @@
...
+from envgenehelper.business_helper import get_namespace_role, get_namespaces, NamespaceRole
 from resource_profiles import processResourceProfiles
 from schema_validation import checkEnvSpecificParametersBySchema
 from cloud_passport import process_cloud_passport
-from pathlib import Path
 
 # const
 GENERATED_HEADER = "The contents of this file is generated from template artifact: %s.\nContents will be overwritten by next generation.\nPlease modify this contents only for development purposes or as workaround."
 
+TEMPLATE_DIR_PATTERN = re.compile(r'/from_(\w+_)?template/')
```
Comment (miyamuraga) [inline scripts/build_env/build_env.py]: move to helper

File: scripts/build_env/build_env.py [hunk 15]
```diff
@@ -5,15 +5,23 @@
...
+
+
+def is_from_template_dir(file_path: str) -> bool:
```
Comment (miyamuraga) [inline scripts/build_env/build_env.py]: move to helper

File: scripts/build_env/build_env.py [hunk 16]
```diff
@@ -184,11 +214,11 @@ def sort_paramsets_with_same_name(entries: list[dict]) -> list[dict]:
     # Lower sort keys are processed first, later values override earlier ones
     def sort_key(e):
```
Comment (miyamuraga) [inline scripts/build_env/build_env.py:198]: pls update tests TestSortParamsetsWithSameName according this changes

File: test_data/test_environments/bgd-cluster/bgd-ns-artifacts-env/Inventory/env_definition.yml [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+  name: "bgd-ns-artifacts"
+  bgNsArtifacts:
+    origin: "bgd-ns-artifacts"
+    peer: "bgd-ns-artifacts"
+  additionalTemplateVariables: {}
+  sharedTemplateVariables: []
+  envSpecificParamsets: {}
+  envSpecificTechnicalParamsets: {}
+  sharedMasterCredentialFiles: []
+  artifact: "bgd-ns-artifacts"
```
Comment (miyamuraga) [inline test_data/test_environments/bgd-cluster/bgd-ns-artifacts-env/Inventory/env_definition.yml]: incorrect example of data, app:ver must be

File: test_data/test_environments/bgd-cluster/bgd-ns-artifacts-env/Inventory/env_definition.yml [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+generatedVersions:
+  generateEnvironmentLatestVersion: "bgd-ns-artifacts:bgd-ns-artifacts"
```
Comment (miyamuraga) [inline test_data/test_environments/bgd-cluster/bgd-ns-artifacts-env/Inventory/env_definition.yml:20]: incorrect example of data, app:ver must be

File: scripts/build_env/tests/env-build/test_render_envs.py [hunk 19]
```diff
@@ -29,16 +29,19 @@ def change_test_dir(self, monkeypatch):
```
Comment (miyamuraga) [inline scripts/build_env/tests/env-build/test_render_envs.py:31]: test data added for env bgd-ns-artifacts, but it is not used in tests so this [code ](https://github.com/Netcracker/qubership-envgene/pull/987/changes#diff-834bc35983abe37fc5bebd98056911865ea42efda477acee0f4a173242e58089R134) is not checked.

File: scripts/build_env/tests/env-build/test_render_envs.py [hunk 20]
```diff
@@ -29,16 +29,19 @@ def change_test_dir(self, monkeypatch):
 
     @pytest.mark.parametrize("cluster_name, env_name, version", test_data)
     def test_render_envs(self, cluster_name, env_name, version):
-        g_templates_dir = str((self.test_data_dir / "test_templates").resolve())
+        g_templates_dirs = {
+            'common': str((self.test_data_dir / "test_templates").resolve())
```
Comment (miyamuraga) [inline scripts/build_env/tests/env-build/test_render_envs.py]: add tests for peer and origin td also

File: scripts/build_env/render_config_env.py [hunk 21]
```diff
@@ -104,16 +107,36 @@ def generate_config(self):
...
-
+            return None
         env_template = openYaml(filePath=env_template_path, safe_load=True)
+        logger.info(f"Loaded env_template from {env_template_path}")
+        return env_template
+
+    def set_env_templates(self):
+        env_template_name = self.ctx.current_env["env_template"]
+
+        env_template_path_stem = f'{self.ctx.templates_dir}/env_templates/{env_template_name}'
```
Comment (miyamuraga) [inline scripts/build_env/render_config_env.py]: used only in error message, no need to create new var

File: scripts/build_env/render_config_env.py [hunk 22]
```diff
@@ -104,16 +107,36 @@ def generate_config(self):
         logger.info(f"config = {config}")
         self.ctx.config = config
 
-    def set_env_template(self):
-        env_template_path_stem = f'{self.ctx.templates_dir}/env_templates/{self.ctx.current_env["env_template"]}'
+    def find_env_template_in_dir(self, template_dir, env_template_name):
+        if not template_dir:
+            return None
+        env_template_path_stem = f'{template_dir}/env_templates/{env_template_name}'
         env_template_path = next(iter(find_all_yaml_files_by_stem(env_template_path_stem)), None)
```
Comment (miyamuraga) [inline scripts/build_env/render_config_env.py]: use new method and rename stem pls to something clearer

File: scripts/build_env/env_template/process_env_template.py [hunk 23]
```diff
@@ -157,27 +152,60 @@ def resolve_artifact_old_logic(env_definition: dict, template_dest: str) -> str:
...
-    unpack_archive(ARTIFACT_DEST, template_dest)
+    artifact_dest = tempfile.mkstemp(suffix='.zip')[1]
+    artifact.download(template_url, artifact_dest, cred)
+    unpack_archive(artifact_dest, template_dest)
     return resolved_version
 
 
-def process_env_template() -> str:
+def is_valid_appver(appver: list[str]) -> bool:
+    return len(appver) >= 2 and bool(appver[0]) and bool(appver[1])
```
Comment (miyamuraga) [inline scripts/build_env/env_template/process_env_template.py]: move to helper

File: scripts/build_env/env_template/process_env_template.py [hunk 24]
```diff
@@ -157,27 +152,60 @@ def resolve_artifact_old_logic(env_definition: dict, template_dest: str) -> str:
...
     project_dir = getenv_with_error('CI_PROJECT_DIR')
-    template_dest = f"{project_dir}/tmp"
-    cluster = getenv_with_error("CLUSTER_NAME")
-    environment = getenv_with_error("ENVIRONMENT_NAME")
-    env_dir = Path(f"{project_dir}/environments/{cluster}/{environment}")
-    env_definition = getEnvDefinition(env_dir)
+    cred_config = render_creds()
 
-    check_dir_exist_and_create(template_dest)
+    for key, appver in appvers.items():
```
Comment (miyamuraga) [inline scripts/build_env/env_template/process_env_template.py]: key is bad name

File: scripts/build_env/env_template/process_env_template.py [hunk 25]
```diff
@@ -157,27 +152,60 @@ def resolve_artifact_old_logic(env_definition: dict, template_dest: str) -> str:
...
+        else:
+            template_dest = f'{project_dir}/tmp/{key}'
 
-    if 'artifact' in env_definition.get('envTemplate', {}):
-        logger.info("Use template resolving new logic")
-        return resolve_artifact_new_logic(env_definition, template_dest)
-    else:
-        logger.info("Use template resolving old logic")
-        return resolve_artifact_old_logic(env_definition, template_dest)
+        if not is_valid_appver(appver):
```
Comment (miyamuraga) [inline scripts/build_env/env_template/process_env_template.py]: why complicate and add choosing method by is_valid_appver ?

File: scripts/build_env/env_template/process_env_template.py [hunk 26]
```diff
@@ -157,27 +152,60 @@ def resolve_artifact_old_logic(env_definition: dict, template_dest: str) -> str:
...
+            if not key == "common":
```
Comment (miyamuraga) [inline scripts/build_env/env_template/process_env_template.py]: more readable just !=

File: scripts/build_env/render_config_env.py [hunk 27]
```diff
@@ -288,30 +335,43 @@ def generate_bgd_file(self):
...
+            role = get_namespace_role(ns_name, bgd) if ns_name else NamespaceRole.COMMON
+
+            role_templates_dir = self._get_template_dir_for_role(role)
+            role_env_template = self._get_env_template_for_role(role)
+
+            effective_ns = ns
+            effective_template_path = ns_template_path
+
+            if role != NamespaceRole.COMMON and role_env_template is not self.ctx.current_env_template:
+                role_ns_config = self._find_ns_config_by_name(role_env_template, ns_name, role_templates_dir)
```
Comment (miyamuraga) [inline scripts/build_env/render_config_env.py:388]: there are no tests related to this condition

**Review by miyamuraga — CHANGES_REQUESTED** [review]

File: scripts/build_env/main.py (hunk 4 above)
Comment (GlimmerCape) [reply scripts/build_env/main.py]: templates_dir is needed in case if user has this macro in their templates https://github.com/Netcracker/qubership-envgene/blob/e7cc8e1538694368bc9b346b5b702a360ef75af1/docs/template-macros.md#templates_dir

File: scripts/build_env/build_env.py (hunk 11 above)
Comment (GlimmerCape) [reply scripts/build_env/build_env.py]: Can you clarify?

File: scripts/build_env/build_env.py (hunk 16 above)
Comment (GlimmerCape) [reply scripts/build_env/build_env.py:198]: You mean sort_paramsets_with_same_name() function to match new function and add new test with origin/peer paramsets? I think it's strange that we essentially use duplicate of function we are testing for preparing test data in tests of the function

File: test_data/test_environments/bgd-cluster/bgd-ns-artifacts-env/Inventory/env_definition.yml (hunk 18 above)
Comment (miyamuraga) [reply test_data/test_environments/bgd-cluster/bgd-ns-artifacts-env/Inventory/env_definition.yml:20]: bgd-ns-artifacts does not look like valid version, app:ver must be. e.g bgd-ns-artifacts: release-1.0.0. also artifact can be snapshot, generateEnvironmentLatestVersion has timestamp. please provide valid examples

File: scripts/build_env/build_env.py [hunk 28]
```diff
@@ -26,25 +25,39 @@ def find_namespaces(dir):
...
-        filePath = str(f)
-        # envSpecific = false will be update later during templates parsing
-        key = extractNameFromFile(filePath)
-        if key in dict:
-            dict[extractNameFromFile(filePath)].append({"filePath": filePath, "envSpecific": False})
-        else:
-            dict[extractNameFromFile(filePath)] = [{"filePath": filePath, "envSpecific": False}]
-    return dict
+def _get_excluded_dirs_for_role(role: NamespaceRole, origin_template_exists: bool, peer_template_exists: bool) -> list:
+    common_dir = 'from_template'
```
Comment (miyamuraga) [inline scripts/build_env/build_env.py:29]: put them directly into **global constants** in helper, in any case, someone else will use them either in tests or in logic

**Review by miyamuraga — CHANGES_REQUESTED** [review]
